### PR TITLE
chore: refactor reconcileEphemeralMetadata to be more concise

### PR DESCRIPTION
**Pull request by zhaque44**
**Changes made:**

The refactored function behaves exactly the same, but it's more concise. The error handling is slightly better and the switch statement makes the code more idiomatic in Go. There is a unit test there to test the logic now

Checklist:

* [ ] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-rollouts/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this is a chore.
* [x] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/) with a list of types and scopes found [here](https://github.com/argoproj/argo-rollouts/blob/master/.github/workflows/pr-title-check.yml), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  
* [x] I've signed my commits with [DCO](https://github.com/argoproj/argoproj)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My builds are green. Try syncing with master if they are not. 
* [ ] My organization is added to [USERS.md](https://github.com/argoproj/argo-rollouts/blob/master/USERS.md).

### CODE REVIEW

**Comment by zhaque44** [comment]
@zachaller question about this PR, it’s saying coverage is 0.03 less from 81.74 to 81.71. I ran the tests in epehemeral tests and they all passed

File: rollout/ephemeralmetadata.go [hunk 1]
```diff
@@ -14,43 +14,46 @@ import (
...
-		}
-		// sync stable metadata to the stable rs
-		err = c.syncEphemeralMetadata(ctx, c.stableRS, stableMetadata)
-		if err != nil {
-			return err
+		// If the rollout is not fully completed, we are in an upgrading state where newRS is a canary.
+		// Sync the new metadata to the newRS.
+		err = c.syncEphemeralMetadata(ctx, c.newRS, newMetadata)
+		// If there was no error syncing the new metadata, sync the stable metadata to the stableRS.
+		if err == nil {
```
Comment (meeech) [inline rollout/ephemeralmetadata.go:41]: I think this is less clear. I would keep the early err returns thoughtout the ifs...

Comment (meeech) [reply rollout/ephemeralmetadata.go:41]: Truth is, if you looking to tighten it up/ lose some indentation, you could keep the return in the if, and then lose the wrapping else entirely

**Comment by zhaque44** [comment]
@meeech @zachaller can you guys kick off the build again. Appreciate it

File: rollout/ephemeralmetadata.go [hunk 2]
```diff
@@ -27,21 +27,19 @@ func (c *rolloutContext) reconcileEphemeralMetadata() error {
...
-		if err != nil {
-			return err
-		}
-		// sync stable metadata to the stable rs
-		err = c.syncEphemeralMetadata(ctx, c.stableRS, stableMetadata)
-		if err != nil {
-			return err
-		}
+		return c.syncEphemeralMetadata(ctx, c.newRS, stableMetadata)
+	}
```
Comment (meeech) [inline rollout/ephemeralmetadata.go:31]: you have changed the behaviour here with this early return. the for loop never happens down below.

Comment (zhaque44) [reply rollout/ephemeralmetadata.go:31]: Thanks so much for your help, that was the issue I resolved it.

File: rollout/ephemeralmetadata.go [hunk 3]
```diff
@@ -27,26 +27,24 @@ func (c *rolloutContext) reconcileEphemeralMetadata() error {
...
-		err := c.syncEphemeralMetadata(ctx, c.newRS, newMetadata)
-		if err != nil {
-			return err
-		}
-		// sync stable metadata to the stable rs
-		err = c.syncEphemeralMetadata(ctx, c.stableRS, stableMetadata)
-		if err != nil {
-			return err
-		}
+		return c.syncEphemeralMetadata(ctx, c.newRS, stableMetadata)
```
Comment (meeech) [inline rollout/ephemeralmetadata.go:30]: The mistake here is you are short circuiting, missing the `for` loop on line -48/+46 which would execute under the old code.

**Pull request closed without merging**